### PR TITLE
[uni fix] connect and switch network (network selector)

### LINKED ISSUE #340: 2440/uni merge: request to switch network is not sent after a wallet connection
Related to #2536

1. Open the app
2. Select Rinkeby network (or any other)
3. Make sure that Rinkeby is not selected in the MM (f.e. GC)
4. Connect to the MM wallet

**AR**: the app is connected to GC, no request to switch network to Rinkeby

**ER**: request to change a network to Rinkeby is sent to MM right after conneciton

> Original issue https://github.com/gnosis/cowswap/issues/2627 by @elena-zh

### LINKED ISSUE #341: 2440/uni merge: ETH is displayed in the Wrap my ETH section in Gnosis chain
Related to #2536 
1. Connect to a wallet in GC
2. Select xDAI token as a sell token, an amount to sell
3. select a buy token

**AR**: Wrap my ETH section appears with ETH token

**ER**: Wrap my xDAI section appears with xDAI token

> Original issue https://github.com/gnosis/cowswap/issues/2629 by @elena-zh

### LINKED ISSUE #352: 2440/uni merge: remove 'in the dropdown menu' from the Wrong Network message
Related to #2536

1. Open the app,
2. Connect to an unsupported network
3. Press on the 'Wrong Network' button

**AR**: Message says: 'Please connect to a supported network **in the dropdown menu** or in your wallet.'
But when  select a network in the dropdown, nothing is changed: no request to change a network is sent to the connected wallet.

**ER**: it would be nice to remove 'in the dropdown menu' from the Wrong Network message
OR change its text to the currently used: 

> Original issue https://github.com/gnosis/cowswap/issues/2645 by @elena-zh

### PR DESCRIPTION
**Pull request by W3stside**
# Summary
This is a 3FER!

Closes #340, closes #352, and closes #341 + 2 minor other broken things

Testing
- disconnect wallet and switch wallet to rinkeby
- using network selector connect to RINKEBY or GCHAIN
- should toggle wallet, switch to network
- check url
- switch url chain id

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @W3stside , great job!

I have found several nitpicks and tricky issues, but let me know please if I need to report a separate issue for them: 
1. Related to #352: 'in the dropdown menu' is still displayed in the warning message. However I noticed the changes logic that selecting a network in the dropdown calls this modal. But actually, it does not resolve the issue itself: a network is not changed in the wallet. So I still vote for removing this phrase from the modal. 
2. Network selector is enabled when connected to WC wallets (as an example, to Safe). So when I try to change a network, I wee this warning message, that seems to me illogical as Safe wallet does not support networks changing. So can we disable this dropdown when connected to SC wallets at least?
Screenshot_2

3. When I connect to MM using WC in a mobile device, I'm able to change networks using dropdown. But after approving this request, I can face the message to open another app, that navigates me to the MM in the Appstore. 
See the video (in this video you can also see that I can't open the dropdown for a while after networks changing): https://drive.google.com/file/d/1jR5AuVWL0O-PMRuIPdQod5uc7AGR-Hm4/view?usp=sharing

Thanks!

File: src/custom/constants/tokens/tokensMod.ts [hunk 1]
```diff
@@ -313,6 +300,17 @@ export function nativeOnChain(chainId: number): NativeCurrency {
...
 
+export class GpEther extends Ether {
+  public get wrapped(): Token {
+    if (this.chainId in WRAPPED_NATIVE_CURRENCY) return WRAPPED_NATIVE_CURRENCY[this.chainId]
+    throw new Error('Unsupported chain ID')
+  }
+
+  private static _cachedExtendedEther: { [chainId: number]: NativeCurrency } = {}
+
+  public static onChain = nativeOnChain
```
Comment (alfetopito) [inline src/custom/constants/tokens/tokensMod.ts:311]: Is this change related to the network selector?

File: src/custom/hooks/useApproveCallback/useApproveCallbackMod.ts [hunk 2]
```diff
@@ -68,31 +68,6 @@ export function useApproveCallback({
   const pendingApproval = useHasPendingApproval(token?.address, spender)
   const spenderCurrency = useCurrency(spender)
 
-  // TODO: Nice to have, can be deleted
-  // eslint-disable-next-line no-lone-blocks
```
Comment (alfetopito) [inline src/custom/hooks/useApproveCallback/useApproveCallbackMod.ts:72]: THANK YOU!

File: src/custom/hooks/useChangeNetworks.ts [hunk 3]
```diff
@@ -1,82 +1,113 @@
...
-type ChangeNetworksParams = Pick<ReturnType<typeof useActiveWeb3React>, 'chainId' | 'library'>
+type ChangeNetworksParams = Pick<ReturnType<typeof useActiveWeb3React>, 'account' | 'chainId' | 'library'>
+export type ChainSwitchCallbackOptions = { skipWalletToggle: boolean; skipToggle: boolean }
 
 /**
  * Hook extracted from Header/NetworkSelector component pretty much verbatim
  *
  * @param chainId
  * @param library
  */
```
Comment (alfetopito) [inline src/custom/hooks/useChangeNetworks.ts:24]: Can you update the doc string as well?

Feel free to remove the params bit, auto generated stuff

File: src/custom/hooks/useChangeNetworks.ts [hunk 4]
```diff
@@ -1,82 +1,113 @@
...
   const dispatch = useAppDispatch()
 
+  const toggleWalletModal = useWalletModalToggle() // MOD
+  const [queuedNetworkSwitch, setQueuedNetworkSwitch] = useState<null | number>(null) // MOD
+  // MOD: get supported chain and check unsupported
+  const chainId = useMemo(() => {
+    const chainId = supportedChainId(preChainId)
+
+    return chainId
+  }, [preChainId])
```
Comment (alfetopito) [inline src/custom/hooks/useChangeNetworks.ts:45]: Nitpick: refactoring

```suggestion
  const chainId = useMemo(() => supportedChainId(preChainId), [preChainId])
```

File: src/custom/pages/App/index.tsx [hunk 5]
```diff
@@ -89,9 +88,6 @@ function createRedirectExternal(url: string) {
 const Loading = <LoadingWrapper>Loading...</LoadingWrapper>
 
 export default function App() {
-  // Dealing with empty URL queryParameters
-  useFilterEmptyQueryParams()
```
Comment (alfetopito) [inline src/custom/pages/App/index.tsx:93]: Wouldn't it break anything else in the app?
I mean, it must have had a reason to be added, right?

File: src/custom/constants/tokens/tokensMod.ts (hunk 1 above)
Comment (alfetopito) [reply src/custom/constants/tokens/tokensMod.ts:311]: Disregard, now I see it's related to an issue you mention in the PR description

File: src/custom/hooks/useChangeNetworks.ts (hunk 3 above)
Comment (alfetopito) [reply src/custom/hooks/useChangeNetworks.ts:24]: Remove the `@param` lines, and update the comment, as it's not verbatin the same code as we had in the `Header/NetworkSelector`.
Or remove the whole docstring

**Comment by alfetopito** [comment]
One thing that this PR is breaking is the ability to switch networks without being connected, which was possible before.
I think that's a great feature and we should keep that.

**Comment by W3stside** [comment]
Merging this. We now have https://github.com/cowprotocol/cowswap/issues/412 and should keep it in mind

**Pull request merged**